### PR TITLE
[DO NOT MERGE] Data Usage Document

### PR DESCRIPTION
**Pull request by eharkins**
Auth0 provides multiple locations for storing different types of data associated with authenticating an app’s users. The purpose of this document is to demonstrate the best practices in using these storage mechanisms efficiently and securely and to give you a look at an end-to-end experience through an application connected to both auth0 and an external database.

### CODE REVIEW

**Comment by auth0build** [comment]
Can one of the admins verify this patch?

**Comment by mpaktiti** [comment]
@eharkins Hey Elias, please resize the following images:
- 3-home.png
- 1-login-just-button.png
- 4-settings.png
  Their height is 1334 and it looks a bit too much.

File: articles/tutorials/data-scenarios/index.md [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+url: /tutorials/user-data-storage-scenario
+---
+
+# Auth0 User Data Storage Scenario
+
+Auth0 provides multiple locations for storing different types of data associated with authenticating an app’s users. The purpose of this document is to demonstrate the best practices in using these storage mechanisms efficiently and securely. It also gives you a look at an example of the end-to-end experience of an application using auth0 and an external database. We created an application to illustrate the important distinctions a developer must make when storing their user data with Auth0. For this example case, we started with a basic mobile app for iOS (coded in Swift) from the Auth0 seed project for an [iOS mobile app](/quickstart/native/ios-swift). As a backend for the app, we used the Auth0 seed project for a simple [Node.js API](/quickstart/backend/nodejs). As we discuss the different types of data and the best places to store them, we will continue to use this application as the example case. See the [Mobile + API architecture scenario](/architecture-scenarios/application/mobile-api) from our documentation to give you a visual of how the application is structured.
```
Comment (mpaktiti) [inline articles/tutorials/data-scenarios/index.md]: @eharkins The paragraph looks too long, maybe break it up a little?

File: articles/tutorials/data-scenarios/index.md [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+- [Where do I put my authentication data?](/tutorials/user-data-storage-scenario#where-do-i-put-my-authentication-data-)
+- [Why not put all the app’s data in the Auth0 data store?](/tutorials/user-data-storage-scenario#why-not-put-all-the-app-s-data-in-the-auth0-data-store-)
+- [When should I use the Auth0 data store?](/tutorials/user-data-storage-scenario#when-should-i-use-the-auth0-data-store-)
+- [Review](/tutorials/user-data-storage-scenario#review)
+
+
+## Where do I put my authentication data?
+
+Auth0 has a data store which serves as a way for developers to store data associated with their users’ profiles, beyond the basic information Auth0 uses for authentication, i.e name, email, username, password, and so on. You can also [Authenticate Users with Username and Password using a Custom Database](/connections/database/mysql) to store such information if your needs are different from what the Auth0 data store offers. However, we recommend storing authentication-related data in Auth0’s data store in order to make things less complicated by allowing you to easily manage user data through Auth0’s [dashboard](${uiURL}).
```
Comment (mpaktiti) [inline articles/tutorials/data-scenarios/index.md:25]: @eharkins I'm not sure that it's clear what these _authentication data_ are. Are we talking `user_metadata` and `app_metadata` here? Maybe add an example or even better a note with a link to the user profile structure doc?

File: articles/tutorials/data-scenarios/index.md [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+
+### Performance
+
+Beyond scale, keeping your authentication data separate from any other data in your app is important because the two sets of data are also likely accessed with different frequencies. The Auth0 data store is not optimized to be queried with extremely high frequency. It is better to leave this up to a specialized database service, optimized for fast, large data operations.
+
+### Flexibility 
+
+By using a separate database, your access to your users' authentication data is concentrated in a small chunk of code, and your access to all other data is grouped separately. Your organizational demands may not be met by the Auth0 data store because it is built to accomodate only the user profiles and their Metadata. Certain actions that you might require from a customizable database service are not possible in Auth0 data store. For example, if you wanted to make a specific query like `SELECT users.favGenre, access.roles FROM users, access WHERE users.user_id = access.user_id`. This is something you can't do with in the auth0 data store. Using a separate database will allow you to manage your data however you see fit.
+
```
Comment (mpaktiti) [inline articles/tutorials/data-scenarios/index.md:37]: @eharkins I would replace these paragraphs (Scalability, Performance, Flexibility) with bullets, and make some redactions at the text as well. For example:
- **Scalability**: The Auth0 data store has limited scalability and your app’s data could exceed that limit. Using an external database allows you to keep things simple on the Auth0 side while leaving the heavy database lifting up to a separate database formatted to efficiently store your extra data.
- **Performance**: ...
- **Flexibility**: ...

File: articles/tutorials/data-scenarios/index.md [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+### Example
+
+Here is an example of data that is associated with a user but not with authenticating that user in the app: In the case of our mobile music application, the user’s music needs to be saved, so they can find it easily when they log in again. This data is not required in the process of authenticating the user for the app, but favorite songs and artists are personal data that should be associated with the user. So we would want to store this data in a separate database connected to the backend of our mobile app, instead of in the Auth0 data store. Here is how we did this:
```
Comment (mpaktiti) [inline articles/tutorials/data-scenarios/index.md]: @eharkins This is the first time this "mobile music application" is mentioned in the doc. When I read the sentence "In the case of our mobile music application" it made me feel like I missed something earlier. Rephrase maybe?

Also, I don't think the ":" after the first sentence looks good, I would replace with "."

File: articles/tutorials/data-scenarios/index.md [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+
+Using the user’s unique `user_id` from their Auth0 [user profile](/user-profile/normalized#storing-user-data), we can make sure the data is associated with the user in our separate database in order to query by a unique identifier. Here is an example row from the `songs` table in our Heroku Postgres database:
```
Comment (mpaktiti) [inline articles/tutorials/data-scenarios/index.md]: @eharkins I would replace the first sentence with something more simple, like "We will use the [user_id](/user-profile/normalized#uniquely-identify-users) as the user's unique identifier. "

File: articles/tutorials/data-scenarios/index.md [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+
+| song_id   | songname           | user_id                    |
+| --------- | ------------------ | -------------------------- |
+| "1a"      | "Yellow Submarine" | "google-oauth2|xxxyyy123" |
+
+The Node.js backend authenticates requests to the specific URI associated with getting the user’s personal data from the database. This is accomplished through the validation of a JSON Web Token. [Learn about token based authentication and how to easily implement JWT in your applications.](https://auth0.com/learn/token-based-authentication-made-easy/)
```
Comment (mpaktiti) [inline articles/tutorials/data-scenarios/index.md]: @eharkins This reference is great, but I would add it as a note:

[Learn about token based authentication and how to easily implement JWT in your applications.](https://auth0.com/learn/token-based-authentication-made-easy/)

File: articles/tutorials/data-scenarios/index.md [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+- Plans: Settings that must not be freely changed by the user without confirmation of a subscription. These require the user to pay or provide proof of purchase through the app to alter their in-app experience in some way.
+- External IDs: Used in associating external accounts (not authentication providers) with the identity provider account that authenticated the user through Auth0.
+
+### Example of `app_metadata`
+
+Some data from our music app that would be appropriate to store in `app_metadata` is music streaming subscriptions. Another example is the user’s permission to edit the app’s featured playlists. Both of these are appropriate for Metadata because they are important in authenticating the user and customizing their experience as they are logged in. What makes them appropriate for `app_metadata` instead of `user_metadata` is that they must not be easily changed by the user. We implemented the permissions example with two Auth0 [rules](/rules):
```
Comment (mpaktiti) [inline articles/tutorials/data-scenarios/index.md]: @eharkins I would remove the ":" from the last paragraph.

File: articles/tutorials/data-scenarios/index.md [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+
+Here's a look at how we allowed the user to change their `displayName`:
+![](/media/articles/tutorials/data-scenarios/4-settings.png)
+
+We used the Auth0 Management APIv2 to allow the app’s users to alter their Metadata via GET and PATCH requests: 
+
+[Get users by id.](/api/management/v2#!/Users/get_users_by_id)
+
+[Patch users by id.](/api/management/v2#!/Users/patch_users_by_id)
+
```
Comment (mpaktiti) [inline articles/tutorials/data-scenarios/index.md:239]: @eharkins Add bullets for these two?

**Comment by mpaktiti** [comment]
@eharkins Hey I left some comments in the doc. A couple more are:
- The seed at the end points to your personal github, not auth0 i.e. https://github.com/eharkins/auth0-swift-data-app This should be under `auth0` or `auth0-samples` org. @ntotten which one?
- I think it would be nice to put a diagram at the beginning of this doc, depicting this example's structure, one seems to have to read quite a bit to get the picture.

**Pull request merged**